playback/src/player.rs

```diff
@@ -1305,14 +1305,20 @@ impl PlayerInternal {
                                 // Engineering Society, 60, 399-408.
                                 if self.config.normalisation_method == NormalisationMethod::Dynamic
                                 {
-                                    // steps 1 + 2: half-wave rectification and conversion into dB
-                                    let abs_sample_db = ratio_to_db(sample.abs());
+                                    if !(self.normalisation_integrator > 0.0
+                                        || self.normalisation_peak > 0.0)
+                                    {
+                                        continue;
+                                    }
 
+                                    let ratio = sample.abs();
                                     // Some tracks have samples that are precisely 0.0, but ratio_to_db(0.0)
                                     // returns -inf and gets the peak detector stuck.
-                                    if !abs_sample_db.is_normal() {
+                                    if ratio == 0.0 {
```

Comment:
This is dangerous for two reasons:

1. [Checking floats for equality is hard](https://floating-point-gui.de/errors/comparison/). If you want to check if it's `0.0`, then you'd have to do `if (ratio - 0.0) < f64::EPSILON { ... }`. Clippy used to warn about this before, I'm surprised it doesn't anymore.

2. The actual intention here is to check that the output of `ratio_to_db(x)` below will always be a number other than `NaN`, `inf` or `-inf`. If you want to do this by checking `x` at the input, then you must be sure to cover all invariants that could lead to such an output, including possible undesired samples from a misbehaving decoder. In this case you could do that by changing this into `sample.is_normal()`. You could bail out early, before `sample.abs()`.

Reply:
I'm trying to think of a reason why we would ever get `Nan` when the function's input is `Vec<f64>`?

Reply:
Possible undesired samples from a misbehaving decoder. It's a check that doesn't cost much, yet brings a lot of robustness.